include/swift/Basic/ExternalUnion.h

```diff
@@ -194,7 +194,7 @@ class BasicExternalUnion {
     static_assert(typeIndex != -1, "type not in union");
     assert(index == Index(typeIndex) && "current kind is wrong for access");
 
-    return reinterpret_cast<const T &>(Storage);
```

Comment:
Wouldn't you need this in `getWithoutIndex` as well? And possibly the non-const members too.

Reply:
Only this line trips over the warning when built with top-of-tree llvm+clang. I can change more lines if people want that.

Reply:
That's probably just about details of what's instantiated in what compilations.  Please do change the other places.

But the concept here is good, thanks.

Reply:
Hi @rjmccall – Since I build with ToT llvm+clang, I wanted you or someone with expertise like you to confirm that this warning is valid and not a recent bug in llvm+clang. Thanks!

Reply:
Oh, I think it's a bug in Clang, yes.  @zygoloid could say for sure.

Since the result type is a reference type, the array-to-pointer conversion on the operand is suppressed, so it's an l-value of type `const char[72]`.  L-value `reinterpret_cast` is defined to be equivalent to doing the corresponding pointer cast, including (per [expr.const.cast]p8) for deciding whether they cast away qualifiers.  Applying that pointer-ization, then, [expr.const.cast]p7 says that we are casting away constness if there is a pair of cv-decompositions of `const char (*)[72]` and `Expr * const *` such that there is no qualification conversion of `const char (*)[72]` to the type formed by replacing the cv-signature of its decomposition with the cv-signature of the corresponding decomposition of `Expr * const *`.  Each of these types has two cv-decompositions.  The source type decomposes with `n=1` (`* {const} U=const char[72]`, cv-signature of `{const}`) and `n=2` (`* {const} [72] {const} U=char`, cv-signature of `{const},{const}`).  The result type decomposes with `n=1` (`* {const} U=Expr*`, cv-signature of `{const}`) and `n=2` (`* {const} * {} U=Expr`, cv-signature of `{const},{}`).  So the question is whether there are qualification-conversions from `const char (*)[72]` to both `* {const} U=const char[72]` (i.e. `const char (*)[72]`) and `* {const} [72] {} U=char` (i.e. `const char (*)[72]`), which of course there are.  Clang is probably getting confused by the way that the array qualifiers collapse when actually building the type in the second pair.